### PR TITLE
Poc/174 store basis data for wholesale

### PR DESCRIPTION
**Pull request by mknic**
## Description

Use the same concepts for storing results when it comes to storing basis data for aggregation and wholesale.

- Use dictionary to contain basis data
- Rewrite `post processor` to take the dictionary and store data for each <key,value> pair in the dictionary
- Prepare for storing basis data for wholesale process. Depends on #167, #168 and #169

When all dependent issues has been solved a follow up PR will be created to make sure basis data are stored correctly for the wholesale process.

## References

* #174

### CODE REVIEW

File: source/databricks/geh_stream/aggregation_utils/trigger_base_arguments.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def trigger_base_arguments():
     p.add('--grid-area', type=str, required=False, help='Run aggregation for specific grid areas format is { "areas": ["123","234"]}. If none is specifed. All grid areas are calculated')
     p.add('--process-type', type=str, required=True, help='D03 (Aggregation) or D04 (Balance fixing) '),
     p.add('--result-url', type=str, required=True, help="The target url to post result json"),
-    p.add('--result-id', type=str, required=True, help="Postback id that will be added to header"),
+    p.add('--result-id', type=str, required=True, help="Postback id that will be added to header. The id is unique"),
     p.add('--persist-source-dataframe', type=bool, required=False, default=False)
     p.add('--persist-source-dataframe-location', type=str, required=False, default="delta/basis-data/")
```
Comment (ZavezX) [inline source/databricks/geh_stream/aggregation_utils/trigger_base_arguments.py:33]: We don't save the basis data in delta format, would a better naming for the basis data path be relevant?

File: source/databricks/geh_stream/shared/services/postprocessor.py [hunk 2]
```diff
@@ -40,16 +40,17 @@ def do_post_processing(self, args, results, now_path_string):
...
 
-    def store_basis_data(self, args, filtered, now_path_string):
+    def store_basis_data(self, args, snapshot_data):
+        snapshot_base = f"{args.persist_source_dataframe_location}{args.result_id}"
 
-        if args.persist_source_dataframe:
-            snapshot_path = f"abfss://{args.data_storage_container_name}@{args.data_storage_account_name}.dfs.core.windows.net/{args.persist_source_dataframe_location}/{now_path_string}"
+        for key, value in snapshot_data.items():
+            path = f"{snapshot_base}/{key}"
+            print(path)
```
Comment (ZavezX) [inline source/databricks/geh_stream/shared/services/postprocessor.py:48]: Print can be removed.

File: source/databricks/geh_stream/shared/services/postprocessor.py [hunk 3]
```diff
@@ -40,16 +40,17 @@ def do_post_processing(self, args, results, now_path_string):
...
+            snapshot_path = f"abfss://{args.data_storage_container_name}@{args.data_storage_account_name}.dfs.core.windows.net/{path}"
```
Comment (ZavezX) [inline source/databricks/geh_stream/shared/services/postprocessor.py:49]: We use this formatting of a path many places in the code. This is a strong candidate for a shared function that can return this formatted path.

File: source/databricks/geh_stream/shared/services/postprocessor.py [hunk 4]
```diff
@@ -40,16 +40,17 @@ def do_post_processing(self, args, results, now_path_string):
                 .format('json').save(result_path)
             # self.coordinator_service.notify_coordinator(path) # TODO
 
-    def store_basis_data(self, args, filtered, now_path_string):
+    def store_basis_data(self, args, snapshot_data):
+        snapshot_base = f"{args.persist_source_dataframe_location}{args.result_id}"
 
-        if args.persist_source_dataframe:
-            snapshot_path = f"abfss://{args.data_storage_container_name}@{args.data_storage_account_name}.dfs.core.windows.net/{args.persist_source_dataframe_location}/{now_path_string}"
+        for key, value in snapshot_data.items():
```
Comment (ZavezX) [inline source/databricks/geh_stream/shared/services/postprocessor.py:46]: Would be more descriptive to use `key, dataframe`.

File: source/databricks/geh_stream/shared/services/postprocessor.py [hunk 5]
```diff
@@ -40,16 +40,17 @@ def do_post_processing(self, args, results, now_path_string):
                 .format('json').save(result_path)
             # self.coordinator_service.notify_coordinator(path) # TODO
 
-    def store_basis_data(self, args, filtered, now_path_string):
+    def store_basis_data(self, args, snapshot_data):
```
Comment (ZavezX) [inline source/databricks/geh_stream/shared/services/postprocessor.py:43]: Would be nice to add types to the function parameters.

**Review by ZavezX — CHANGES_REQUESTED** [review]

File: source/databricks/geh_stream/aggregation_utils/aggregators/aggregation_initializer.py [hunk 6]
```diff
@@ -77,7 +77,7 @@ def get_translated_grid_loss_sys_corr(args, spark):
     return load_grid_loss_sys_corr(args, spark)
 
 
-def get_time_series_dataframe(args, areas, spark):
+def get_time_series_dataframe(args, areas, spark, snapshot_data):
```
[the remark this answers is not in the document]
Comment (mknic) [reply source/databricks/geh_stream/aggregation_utils/aggregators/aggregation_initializer.py:80]: Very good input. I will make that change.

File: source/databricks/aggregation-jobs/wholesale_trigger.py [hunk 7]
```diff
@@ -41,22 +42,38 @@
...
 
+# Dictionary containing raw data frames
+snapshot_data = {}
+
 # Load raw data frames based on date and grid area filters
 time_series = load_time_series(args, spark, grid_areas)
+snapshot_data["time_series_df"] = time_series
 
 metering_points = load_metering_points(args, spark, grid_areas)
+snapshot_data["metering_points_df"] = metering_points
```
[the remark this answers is not in the document]
Comment (mknic) [reply source/databricks/aggregation-jobs/wholesale_trigger.py:53]: I need you to elaborate on this. I'm not sure I understand what you mean.

File: source/databricks/geh_stream/aggregation_utils/trigger_base_arguments.py (hunk 1 above)
Comment (mknic) [reply source/databricks/geh_stream/aggregation_utils/trigger_base_arguments.py:33]: Agree but as we talked about yesterday I will create an issue to do the renaming. We have other arguments we want to rename and I would think we could save some time if we take the discussion about naming upfront before the implementation is started.

File: source/databricks/geh_stream/shared/services/postprocessor.py (hunk 5 above)
Comment (mknic) [reply source/databricks/geh_stream/shared/services/postprocessor.py:43]: Good idea!

File: source/databricks/aggregation-jobs/wholesale_trigger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZavezX) [reply source/databricks/aggregation-jobs/wholesale_trigger.py:53]: The `snapshot_data` is a dictionary of `key, dataframe`. That means we can get the dataframe back by it's key from the dictionary. Therefor we don't need the variables for each dataframe as we already have them in the dictionary.

File: source/databricks/geh_stream/shared/services/postprocessor.py (hunk 5 above)
Comment (mknic) [reply source/databricks/geh_stream/shared/services/postprocessor.py:43]: I can see our codebase lacks of this convention having types in function parameters. Maybe we need an issue to handle it if we want it that way?

**Review by ZavezX — APPROVED** [review]
Looks good.

**Review by johevemi — APPROVED** [review]

**Pull request merged**